From 75f655a15dcf53c0b796d98918c6e88f5e86fb75 Mon Sep 17 00:00:00 2001
From: Bob Paauwe <bob.j.paauwe@intel.com>
Date: Mon, 27 Jul 2020 14:23:24 -0700
Subject: [PATCH] drm/i915: Add Add Wa_14010685332

The WA specifies that we need to toggle a SDE chicken bit on and then
off as the final step in preparation for runtime suspend.

    Bspec: 33450
    Bspec: 8402

However, something is happening after we toggle the bit that causes
the WA to be invalidated. Instead, lets try setting the bit on suspend
and clearing it on resume.

Signed-off-by: Bob Paauwe <bob.j.paauwe@intel.com>
---
 .../gpu/drm/i915/display/intel_display_power.c  | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/display/intel_display_power.c b/drivers/gpu/drm/i915/display/intel_display_power.c
index 9ea51b6c769b..7208c3f70482 100644
--- a/drivers/gpu/drm/i915/display/intel_display_power.c
+++ b/drivers/gpu/drm/i915/display/intel_display_power.c
@@ -5452,17 +5452,30 @@ static void intel_power_domains_verify_state(struct drm_i915_private *i915)
 
 void intel_display_power_suspend_late(struct drm_i915_private *i915)
 {
-	if (INTEL_GEN(i915) >= 11 || IS_GEN9_LP(i915))
+	struct intel_uncore *uncore = &i915->uncore;
+
+	if (INTEL_GEN(i915) >= 11 || IS_GEN9_LP(i915)) {
 		bxt_enable_dc9(i915);
-	else if (IS_HASWELL(i915) || IS_BROADWELL(i915))
+		/* Wa_14010685332:icl+ */
+		if (INTEL_GEN(i915) >= 11)
+			intel_uncore_rmw(uncore, SOUTH_CHICKEN1,
+				 SBCLK_RUN_REFCLK_DIS, SBCLK_RUN_REFCLK_DIS);
+	} else if (IS_HASWELL(i915) || IS_BROADWELL(i915)) {
 		hsw_enable_pc8(i915);
+	}
 }
 
 void intel_display_power_resume_early(struct drm_i915_private *i915)
 {
+	struct intel_uncore *uncore = &i915->uncore;
+
 	if (INTEL_GEN(i915) >= 11 || IS_GEN9_LP(i915)) {
 		gen9_sanitize_dc_state(i915);
 		bxt_disable_dc9(i915);
+		/* Wa_14010685332:icl+ */
+		if (INTEL_GEN(i915) >= 11)
+			intel_uncore_rmw(uncore, SOUTH_CHICKEN1,
+					 SBCLK_RUN_REFCLK_DIS, 0);
 	} else if (IS_HASWELL(i915) || IS_BROADWELL(i915)) {
 		hsw_disable_pc8(i915);
 	}
}
-- 
2.21.0

